Incorporate AHRS acc2q into zscilib as zsl_quat_from_accel

deganii opened the pull request:
Currently, the zscilib's Madgwick orientation filter starts with an orientation of <1, 0, 0, 0> regardless of the physical IMU's orientation. 

The filter then begins the slow/painful process of converging to the correct orientation which can take nearly a minute. For an IMU running at 120Hz, the worst-case Madgwick convergence time is about 47 seconds (β * Dt = 0.033 * 1/120Hz). This is far too long for a warm-up; a Madgwick filter that has not yet converged will output spurious and confusing orientations and result in a bad user experience. 

A python implementation of Magwick, called AHRS, solves this bootstrapping problem with a non-iterative algorithm called acc2q (acceleration => quat) 

AHRS Implementation:
https://github.com/Mayitzin/ahrs/blob/595ab9877c77f5960e39b352c0ef551d2d2efcb9/ahrs/common/orientation.py#L970
Reference:
([Trimpe et al] https://idsc.ethz.ch/content/dam/ethz/special-interest/mavt/dynamic-systems-n-control/idsc-dam/Research_DAndrea/Balancing%20Cube/ICRA10_1597_web.pdf

This algorithm that relies only on the current accelerometer sample. The integration of the acc2q algorithm is highlighted in the below flow diagram (dotted red block). 

![diagram](https://github.com/user-attachments/assets/6a32fe55-2233-4b40-bec4-85e4af0750a3)

Computing q0 and feeding this into Madgwick results in a very reasonable q0 orientation and a much better overall experience.

deganii committed:
Add zsl_quat_from_accel implementation to quaternions.c

Add zsl_quat_from_accel header to quaternions.h

Incorporate AHRS acc2q into zscilib as zsl_quat_from_accel

Currently, the zscilib's Madgwick orientation filter starts with an orientation of <1, 0, 0, 0> regardless of the physical IMU's orientation.

The filter then begins the slow/painful process of converging to the correct orientation which can take literally minutes. For an IMU running at 120Hz, the worst-case Madgwick convergence time is about 47 seconds (β * Dt = 0.033 * 1/120Hz). This is far too long for a warm-up; a Madgwick filter that has not yet converged will output spurious and confusing orientations and result in a bad user experience.

A python implementation of Magwick, called AHRS, solves this bootstrapping problem with a non-iterative algorithm called acc2q (acceleration => quat)

AHRS Implementation:
https://github.com/Mayitzin/ahrs/blob/595ab9877c77f5960e39b352c0ef551d2d2efcb9/ahrs/common/orientation.py#L970
Reference:
([Trimpe et al] https://idsc.ethz.ch/content/dam/ethz/special-interest/mavt/dynamic-systems-n-control/idsc-dam/Research_DAndrea/Balancing%20Cube/ICRA10_1597_web.pdf

This algorithm that relies only on the current accelerometer sample. The integration of the acc2q algorithm is highlighted in the below flow diagram (dotted red block).

![diagram](https://github.com/user-attachments/assets/6a32fe55-2233-4b40-bec4-85e4af0750a3)

Computing q0 and feeing this into Madgwick results in a very reasonable q0 orientation and a much better overall experience.

Signed-off-by: Ismail Degani <deganii@gmail.com>

microbuilder commented:
As per the comment in #62, can you create a PR in Zephyr that references this commit so that it can go through the normal test suite before being merged? Thanks for the changes, though ... much appreciated.

microbuilder commented:
Please also add some details in the commit messages and add a signed-off-by to make sure you get the proper credit for the work.

deganii commented:
Thanks for the review. Please see my updates:

- I've modified to a single commit which has all the details of the fixes made
- I've added the signoff requested
- I've fixed all of the comment / lint issues
I've raised a top-level commit which refers to this pull request in the submanifest optional.yaml :

https://github.com/zephyrproject-rtos/zephyr/pull/78596

kartben commented:
Ping @microbuilder

microbuilder commented:
> Ping @microbuilder 

As per some comments in the discord there are still quite a few issues with this preventing it from being it from being mergable.
